### ISSUE TITLE
RuntimeError: vk::Device::allocateDescriptorSetsUnique: ErrorOutOfPoolMemory

### ISSUE THREAD

**Issue by chaoyivision** [haosulab/SAPIEN-Release#30]
**System:**
 - OS version: Ubuntu 16.04
 - Python version (if applicable): Python 3.7
 - SAPIEN version (pip freeze | grep sapien): sapien-2.0-cp37-cp37m-manylinux2014_x86_64.whl
 - Environment: Desktop

**Describe the bug**
Hi fb, I'd like to use sapien to iterate all 2347 shapes within one python script (like a dataloader). To achieve that, I built one unique engine (=sapien.Engine(0, 0.001, 0.005)), one unique renderer (=sapien.VulkanRenderer(offscreen_only=not self.settings['show_gui'])), and one unique scene(=self.engine.create_scene(config=_scene_config_)), and one unique viewer (Viewer(self.renderer)). 

Therefore, I will (1) create a new camera and load a new object, (2) update viewer's parameters (including x,y,z,r,p,y, and fovy) given the newly created camera, and (3) call **self.viewer.render()** for visualization purpose _at the beginning of each iteration_, and remove them from self.scene _at the end of each iteration_.

For the first 210+ objects, everything works fine - I could successfully iterative among them and pop a viewer window for each. However, the following error showed up. I suppose this error has nothing to do with any corrupted shapes, as (1) when I repeat the script multiple times, it still runs successfully for first 210+ iterations, but could be stopped at different iterations (like 213, 215, etc), (2) I have monitored both my GPU and RAM - both of them look far below the limitation when error poped.

May I have your guidance or comments on this, please? Thanks!

Update: I have tried to re-run the scrips without self.viewer (and self.viewer.render()), i.e., no visualization/GUI now, the same error would pop at **self.camera.take_picture()** around 210+ iterations.

**ErrorTrack**
Traceback (most recent call last):
  File "xx", line xx, in xx
    self.viewer.render()
  File "/home/xx/anaconda3/lib/python3.7/site-packages/sapien/utils/viewer.py", line 1786, in render
    self.info_window,
RuntimeError: vk::Device::allocateDescriptorSetsUnique: ErrorOutOfPoolMemory

**Comment by fbxiang** [comment]
We currently do not delete any render mesh from GPU. In future versions, we plan to provide a function to clear this cache. For now, you can destroy engines and renderers and recreate them to clear resources.

**Comment by chaoyivision** [comment]
Thanks, fb. I'm a bit confused about "destroy engines", as I recalled that each process could only launch one engine (pls correct me if I am wrong about this, thx). Does it mean, for now, I have to shut down the entire python process after the first 210+ iterations? Is there any way to achieve **the iteration among 2347 shapes within one python script** at the current stage?

**Comment by fbxiang** [comment]
I typically use the multiprocessing package, essentially using python to spawn python code in multiple processes.

**Comment by harryzhangOG** [comment]
Hi, I am running into the same problem. How do I destroy the engine? Thanks.

**Comment by fbxiang** [comment]
We have provided a function to clear the resources at https://github.com/haosulab/SAPIEN/blob/master/python/pysapien_content.hpp#L1766. In the release versions, the engine cannot be destroyed. However, I believe in dev branch, we decided that when all references to the engine and renderer are released (set everything related to sapien to None), all resources will be released. You will probably want to open new issues with your specific information since we do not tend to monitor closed issues.
